src/validation.h

```diff
@@ -822,14 +822,41 @@ class ChainstateManager
     //! free the chainstate contents immediately after it finishes validation
     //! to cautiously avoid a case where some other part of the system is still
     //! using this pointer (e.g. net_processing).
+    //!
+    //! Once this pointer is set to a corresponding chainstate, it will not
+    //! be reset until init.cpp:Shutdown(). This means it is safe to acquire
+    //! the contents of this pointer with ::cs_main held, release the lock,
+    //! and then use the reference without concern of it being deconstructed.
+    //!
+    //! This is especially important when, e.g., calling ActivateBestChain()
+    //! on all chainstates because we are not able to hold ::cs_main going into
+    //! that call.
     std::unique_ptr<CChainState> m_ibd_chainstate;
 
     //! A chainstate initialized on the basis of a UTXO snapshot. If this is
     //! non-null, it is always our active chainstate.
+    //!
+    //! Once this pointer is set to a corresponding chainstate, it will not
+    //! be reset until init.cpp:Shutdown(). This means it is safe to acquire
```

Comment:
In commit "protect g_chainman with cs_main" (d88f75e495a46ae474d8ad67faaec13ef1950f5e)

This doesn't explain what happens if you load two snapshots (one after the other). It seems like `ChainstateManager::InitializeChainstate` explicitly allows this so it would be good to add some clarification here or there.

Reply:
I don't think `InitializeChainstate` allows this because of this check: https://github.com/jamesob/bitcoin/blob/8fe8b2e499ceb61c6fd3b73bf1c5dbc5861ccf62/src/validation.cpp#L5164-L5170

I wrote it to not allow the initialization of more than one "kind" of chainstate.